transmogrifier/datacite.py

```diff
@@ -0,0 +1,187 @@
+            resource_types.append(
+                Note(value=resource_type_xml.text, kind="ResourceType")
+            )
+        if "resourceTypeGeneral" in resource_type_xml.attrib:
+            content_types.append(resource_type_xml.attrib["resourceTypeGeneral"])
+    kwargs["notes"] = resource_types
+    kwargs["content_type"] = content_types
+
+    kwargs = {k: v for k, v in kwargs.items() if v}
+    return kwargs
+
+
+def create_instance_list_from_xml(
+    xml_element: Any,
+    xpath: str,
+    instance_class: Any,
+    attribute_mapping_dict: Dict[str, str],
+    value_child_element: Optional[str] = None,
+) -> list[Any]:
+    """
+    Create a list of instances of the specificed class and populate those instances
+    with data from the XML element being parsed.
+    Args:
+        xml_element: The XML element to parse.
+        xpath: The xpath to use for parsing the XML element for values.
+        instance_class: The class used to create the instances that will be populated.
+        attribute_mapping_dict: A dict mapping instance attributes to their source XML
+        attributes. Passed to the populate_instance_attributes_from_xml function.
+        value_child_element_tag: The tag of the child element that contains the value for
+        the instance's value attribute. Passed to the get_instance_value_from_xml
+        function.
+    """
+    instance_list = []
+    xml_elements = xml_element.findall(xpath, NAMESPACES)
+    for xml_element in xml_elements:
+        instance = instance_class(
+            value=get_instance_value_from_xml(xml_element, value_child_element)
+        )
+        instance = populate_instance_attributes_from_xml(
+            instance,
+            xml_element,
+            attribute_mapping_dict,
+        )
+        instance_list.append(instance)
+    return instance_list
+
+
+def create_value_list_from_xml(
+    xml_element: Any,
+    xpath: str,
+) -> list[Any]:
+    """
+    Args:
+        xml_element: The XML element to parse.
+        xpath: The xpath to use for parsing the XML element for values.
+    """
+    value_list = []
+    xml_elements = xml_element.findall(xpath, NAMESPACES)
+    for xml_element in xml_elements:
+        value_list.append(xml_element.text)
+    return value_list
+
+
+def get_instance_value_from_xml(
+    xml_element: Any, child_element_tag: Optional[str] = None
+) -> str:
+    """
+    Get the value for populating an instance's value attribute by parsing an XML element.
+    If no child element is specified, the text attribute of the XML element will be
+    retrieved.
+    Args:
+        xml_element: The XML element to parse.
+        child_element_tag: The tag of the child element that contains the value for
+        the instance's value attribute.
+    """
+    if child_element_tag:
+        value = xml_element.find(child_element_tag, NAMESPACES).text
+    else:
+        value = xml_element.text
+    return value
+
+
+def populate_instance_attributes_from_xml(
```

Comment:
I’m also unsure how to handle type hinting for `populate_instance_attributes_from_xml_attributes` since it accepts and returns different classes like `Contributor` or `Date` and my attempts at doing a Union failed.